src/util/indexing.jl

```diff
@@ -0,0 +1,263 @@
+names[2] -> vals[2] ... and so on.
+"""
+function Indexer{S <: AbstractString, T}(names::Vector{S}, vals::Vector{T})
+    Indexer(convert(Vector{Symbol}, names), vals)
+end
+
+
+"""
+Construct an Indexer that maps a Vector of Symbols (names), to a set of
+values of type T <: Integer.
+
+Provide this constructor a vector of Symbols as names, and a type that is a
+subtype of Integer, and it will construct an index such that
+names[1] -> 1, names[2] -> 2, ..., names[length(names)] -> length(names).
+"""
+function Indexer{T <: Integer}(names::Vector{Symbol}, ::Type{T})
+    return Indexer(names, collect(T(1):T(length(names))))
+end
+
+
+"""
+Construct an Indexer that maps a Vector of Strings (names), to a set of
+values of type T <: Integer.
+
+Provide this constructor a vector of Strings as names, and a type that is a
+subtype of Integer, and it will construct an index such that
+names[1] -> 1, names[2] -> 2, ..., names[length(names)] -> length(names).
+"""
+function Indexer{S <: AbstractString, T <: Integer}(names::Vector{S}, ::Type{T})
+    return Indexer(convert(Vector{Symbol}, names), T)
+end
+
+
+"""
+Construct an Indexer that maps a Vector of Symbols (names), to a set of
+values of type Int, where Int is the default integer encoding on your machine
+(usually Int32 or Int64).
+
+Provide this constructor a vector of Symbols as names, and it will construct
+an index such that names[1] -> 1, names[2] -> 2, ...,
+names[length(names)] -> length(names).
+"""
+function Indexer(names::Vector{Symbol})
+    return Indexer(names, Int)
+end
+
+
+"""
+Construct an Indexer that maps a Vector of Strings (names), to a set of
+values of type Int, where Int is the default integer encoding on you machine
+(usually Int32 or Int64).
+
+Provide this constructor a vector of Strings as names, and it will construct
+an index such that names[1] -> 1, names[2] -> 2, ...,
+names[length(names)] -> length(names).
+"""
+function Indexer{S <: AbstractString}(names::Vector{S})
+    return Indexer(convert(Vector{Symbol}, names), Int)
+end
+
+
+"Make sure all vectors in a symbol are unique."
+function make_unique(names::Vector{Symbol})
+    seen = Set{Symbol}()
+    names = copy(names)
+    dups = Int[]
+    for (i, name) in enumerate(names)
+        in(name, seen) ? push!(dups, i) : push!(seen, name)
+    end
+    for i in dups
+        nm = names[i]
+        k = 1
+        while true
+            newnm = symbol("$(nm)_$k")
+            if !in(newnm, seen)
+                warn("Trying to build an index with duplicate names. Changing $(names[i]) to $(newnm)")
+                names[i] = newnm
+                push!(seen, newnm)
+                break
+            end
+            k += 1
+        end
+    end
+    return names
+end
+
+
+Base.length(x::Indexer) = length(x.names)
+Base.names(x::Indexer) = copy(x.names)
+_names(x::Indexer) = x.names
+Base.copy(x::Indexer) = Indexer(copy(x.lookup), copy(x.names))
+Base.deepcopy(x::Indexer) = copy(x) # all eltypes immutable
```

Comment:
Was a throwback to when the code represented some code I used from DataFrames.jl, Fixed :+1: